src/main/java/org/junitpioneer/jupiter/AbstractEntryBasedExtension.java

```diff
@@ -43,60 +45,56 @@
  * @param <S> The set annotation type.
  */
 abstract class AbstractEntryBasedExtension<K, V, C extends Annotation, S extends Annotation>
-		implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback {
-
-	@Override
-	public void beforeAll(ExtensionContext context) {
-		clearAndSetEntries(context);
-	}
+		implements BeforeEachCallback, AfterEachCallback {
 
 	@Override
 	public void beforeEach(ExtensionContext context) {
-		clearAndSetEntries(context);
+		/*
+		 * We cannot use PioneerAnnotationUtils#findAllEnclosingRepeatableAnnotations(ExtensionContext, Class) or the
+		 * like as clearing and setting might interfere. Therefore, we have to apply the extension from the outermost
+		 * to the innermost ExtensionContext.
+		 */
```

Comment:
Is this comment clear to everyone? Consider the following:

```java
@SetSystemProperty(key = "foo", value = "42")
class MyTests {

    void testA() {
        // ...
    }

    @ClearSystemProperty(key = "foo")
    void testB() {
        // ...
    }

}
```

Note that we first clear and then set entries (here system properties):

https://github.com/junit-pioneer/junit-pioneer/blob/0ef4ca66fe5766b3419a7846104cf6aa1b788c09/src/main/java/org/junitpioneer/jupiter/AbstractEntryBasedExtension.java#L76-L77

(If we first set and then clear, the problem still persist with the inverse example, i.e., `@ClearSystemProperty` on the class and `@SetSystemProperty` on the method.)

Now when the `BeforeEachCallback` is invoked and we use `PioneerAnnotationUtils#findAllEnclosingRepeatableAnnotations(ExtensionContext, Class)`, we first clear `foo` and then set `foo`, which is not what the user wants.

Reply:
`
> 
> Note that we first clear and then set entries (here system properties):
> 

But den quoted docs tell otherway (from outer to inner, so suggest to first set, then clear)

Reply:
We do both, clear and set per `ExtensionContext`. `BeforeEachCallback` from outermost to innermost, `AfterEachCallback` from innermost to outermost.

Reply:
Side note: I meant this from an implementation perspective. This is what the code does:

https://github.com/junit-pioneer/junit-pioneer/blob/0ef4ca66fe5766b3419a7846104cf6aa1b788c09/src/main/java/org/junitpioneer/jupiter/AbstractEntryBasedExtension.java#L76-L77

First clear, then set. From a user perspective, it is, of course, different. For the given example, as we clear and set per extension context:

* `BeforeEachCallback`: from outermost to innermost:
  1. Class-level / `MyTest`:
    * Clear: n/a
    * Set: `foo = 42`
  2. Method-level / `testB()`:
    * Clear: `foo`
    * Set: n/a
* `AfterEachCallback`: from innermost to outermost:
  1. Method-level / `testB()`:
    * Restore `foo = 42`
  2. Class-level / `MyTest`:
    * Restore `foo` to pre-test state